### PR TITLE
download linuxkit instead of building, as it is a basic tool; eliminate `eve-build-<user>` image from most activities

### PR DESCRIPTION
**Pull request by deitch**
This changes the `Makefile` so that linuxkit is downloaded directly from its GitHub release page as a compiled binary, rather than downloading the repo and compiling it. It also eliminates the need for a "linuxkit.version" flag file, but checking the actual linuxkit version against the desired version by running `linuxkit version --short`. As of v1.3.0, linuxkit supports that syntax.

The real benefit, though, is that it eliminates the need for building that `eve-build-<user>` docker image, unless you take an activity that requires it (like `make shell`). Those activities are all developer-specific ones. With this change, it will build that only if you take one of those activities.

`make eve-pillar` is **not** one of those activities, nor is any of the other `make eve-*` ones for packages, or even `make eve`.

This will make getting to the actual build part of `make` commands faster.

Also, perhaps even more importantly, it eliminates the usage of that image from the GitHub Actions pipelines. That means no more pulls of `golang` or similar for that purpose. This will make logs cleaner, and a *lot* fewer pulls of rate-limited images. We were pulling them with every single run of any command to the Makefile. That goes away.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -302,8 +302,10 @@ DOCKER_GO = _() { $(SET_X); mkdir -p $(CURDIR)/.go/src/$${3:-dummy} ; mkdir -p $
...
 LINUXKIT=$(BUILDTOOLS_BIN)/linuxkit
-LINUXKIT_VERSION=e6b0ae05eb3a2b99e84d9ffc03a3a5c9c3e7e371
-LINUXKIT_SOURCE=https://github.com/linuxkit/linuxkit.git
+# linuxkit version. This **must** be a published semver version so it can be downloaded already compiled from
+# the release page at https://github.com/linuxkit/linuxkit/releases
+LINUXKIT_VERSION=v1.3.0
+LINUXKIT_SOURCE=https://github.com/linuxkit/linuxkit
 LINUXKIT_OPTS=$(if $(strip $(EVE_HASH)),--hash) $(EVE_HASH) $(if $(strip $(EVE_REL)),--release) $(EVE_REL)
 LINUXKIT_PKG_TARGET=build
 LINUXKIT_PATCHES_DIR=tools/linuxkit/patches
```
Comment (giggsoff) [inline Makefile:311]: Do we plan to use patches (again?)? As I understand we want to stabilise upstream instead, correct? BTW we can move to that case if required in future.

**Review by giggsoff — COMMENTED** [review]
I can see tons of lines in [workflow](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344334#step:9:903) (something like `tar: Ignoring unknown extended header keyword...`) can you please check?

**Comment by deitch** [comment]
A few things:

First, I do not understand why [the riscv64 PR build failed](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344812?pr=4081). 

Second, it looks like `make eve` still tries to create `eve-build-ubuntu`. See for example [this line](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339343585?pr=4081#step:9:12) (arm64, xen, generic) or [this line](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344567?pr=4081#step:9:10) (amd64, kvm, generic).

I cannot recreate it locally. If I try to run the same `make` command as in the workflow logs, I do **not** get it trying to build the `eve-build-*` image.

It does not build for the pkgs, which is a definite step up. But we should resolve the rest before merging this in.

File: Makefile (hunk 1 above)
Comment (deitch) [reply Makefile:311]: We do not, and we have not for a while. We should try as much as we can to use off-the-shelf upstream components.

Comment (giggsoff) [reply Makefile:311]: If so, let's delete `LINUXKIT_PATCHES_DIR`.
And please take a look onto the workflow log (I mentioned it in the review comment), 40000 of lines looks sub-optimal.

**Comment by deitch** [comment]
> I can see tons of lines in [workflow](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344334#step:9:903) (something like tar: Ignoring unknown extended header keyword...) can you please check?

Yes, I know what it is. linuxkit tracks information on each file that was placed into its tar (rootfs.tar) so it can know efficiently if it needs to redo them when rebuilding. It turns out to make it significantly faster (multiple times) to regenerate. Added in [this PR](https://github.com/linuxkit/linuxkit/pull/4025). This was an efficiency idea from @rouming, which turned out to be really helpful.

It stores the additional info as PAX headers on the files. Not every version of tar recognizes extended PAX headers. Those that do not, usually treat it as a warning, which is convenient. It continues to work backwards-compatibly. 

The messages you are seeing is because we call `tar xf rootfs.tar` in `make sbom`, and since gnu `tar` on the runner does not support those headers, it will give the message.

We probably can (should?) fix it, although unrelated to this PR. We can suppress it with `tar --warning=no-unknown-keyword`. Separate PR for that, methinks.

**Comment by deitch** [comment]
Mind, `tar --warning=` only works on GNU tar, whereas `bsdtar` has no issues with those. Probably need to check tar version first.

**Comment by deitch** [comment]
It looks like the `packages` job in `build.yml` is getting the latest `Makefile` but the `eve` job is not? I added a temporary commit that dumps the Makefile to see.

File: Makefile (hunk 1 above)
Comment (deitch) [reply Makefile:311]: > If so, let's delete LINUXKIT_PATCHES_DIR.

Sure, good idea.

> And please take a look onto the workflow log (I mentioned it in the review comment), 40000 of lines looks sub-optimal

Yeah, separate PR. It has been a problem for a bit. Will clean it up.

Comment (deitch) [reply Makefile]: > If so, let's delete LINUXKIT_PATCHES_DIR.

Done

Comment (deitch) [reply Makefile]: And see #4082 for suppressing those warnings.

**Review by christoph-zededa — APPROVED** [review]

**Comment by deitch** [comment]
> First, I do not understand why [the riscv64 PR build failed](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344812?pr=4081).

This was transient, rerunning it fixed it.

> Second, it looks like make eve still tries to create eve-build-ubuntu. See for example [this line](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339343585?pr=4081#step:9:12) (arm64, xen, generic) or [this line](https://github.com/lf-edge/eve/actions/runs/9896383146/job/27339344567?pr=4081#step:9:10) (amd64, kvm, generic).

This was due to an update necessary in `build.yml` unrelated to this PR. Yash is fixing that one, when it is done, I will rebase this PR on that fix.

**Comment by deitch** [comment]
Yash fixed it, rebased, should be clean now. Will follow closely.

**Comment by deitch** [comment]
I can confirm that the build now does **not** try to build linuxkit or the `eve-build-ubuntu` image! I checked multiple `make` calls in the action run logs.

**Review by rene — APPROVED** [review]

**Review by yash-zededa — APPROVED** [review]

**Pull request merged**